hw/ip/prim/util/primgen.py

```diff
@@ -0,0 +1,383 @@
+    #   parameter_declaration
+    #   | local_parameter_declaration
+    #   | data_type list_of_param_assignments
+    #   | type list_of_type_assignments
+
+    # XXX: Not covering the complete grammar, e.g. `parameter x, y`
+    RE_PARAMS = (
+        r'parameter\s+'
+        r'(?:[a-zA-Z0-9\]\[:\s\$]+\s+)?'  # type
+        r'(?P<name>\w+)'  # name
+        r'(?:\s*=\s*[^,;]+)'  # initial value
+    )
+    re_params = re.compile(RE_PARAMS)
+    parameters = set()
+    for m in re_params.finditer(parameter_port_list):
+        parameters.add(m.group('name'))
+    return parameters
+
+
+def _generate_prim_pkg(gapi):
+    all_prim_cores = _prim_cores(gapi['cores'])
+    techlibs = _techlibs(all_prim_cores)
+
+    techlib_enums = []
+
+    # Insert the required generic library first to ensure it gets enum value 0
+    techlib_enums.append(_enum_name_for_techlib('generic', qualified=False))
+
+    for techlib in techlibs:
+        if techlib == 'generic':
+            # The generic implementation is required and handled separately.
+            continue
+        techlib_enums.append(_enum_name_for_techlib(techlib, qualified=False))
+
+    # Render prim_pkg.sv file
+    print("Creating prim_pkg.sv")
+    prim_pkg_sv_tpl_filepath = os.path.join(os.path.dirname(__file__),
+                                            'primgen', 'prim_pkg.sv.tpl')
+    prim_pkg_sv_tpl = Template(filename=prim_pkg_sv_tpl_filepath)
+
+    prim_pkg_sv = prim_pkg_sv_tpl.render(encoding="utf-8",
+                                         techlib_enums=techlib_enums)
+    with open('prim_pkg.sv', 'w') as f:
+        f.write(prim_pkg_sv)
+
+    # Copy prim_pkg.core (no changes needed)
+    prim_pkg_core_src = os.path.join(os.path.dirname(__file__), 'primgen',
+                                     'prim_pkg.core.tpl')
+    prim_pkg_core_dest = 'prim_pkg.core'
+    shutil.copyfile(prim_pkg_core_src, prim_pkg_core_dest)
+    print("Core file written to %s." % (prim_pkg_core_dest, ))
+
+
+def _instance_sv(prim_name, techlib, parameters):
+    s = "prim_{techlib}_{prim_name} #(\n"
+    s += ", ".join(".{p}({p})".format(p=p) for p in parameters)
+    s += ") u_impl_{techlib} (\n" \
+         "  .*\n" \
+         ");\n"
+    return s.format(prim_name=prim_name, techlib=techlib)
+
+
+def _create_instances(prim_name, techlibs, parameters):
+    """ Build SystemVerilog code instantiating primitives from the techlib """
+    techlibs_wo_generic = [
+        techlib for techlib in techlibs if techlib != 'generic'
+    ]
+    techlibs_generic_last = techlibs_wo_generic + ['generic']
+
+    if not techlibs_wo_generic:
+        # Don't output the if/else blocks if there no alternatives exist.
+        # We still want the generate block to keep hierarchical path names
+        # the stable, even if more than one techlib is found.
+        s = "generate begin : gen_generic\n"
+        s += _instance_sv(prim_name, "generic", parameters) + "\n"
+        s += "end endgenerate"
```

Comment:
this unfortunately causes DC to throw this error:
```
Syntax error on an obsolete Verilog 2001 construct standalone generate block 'gen_generic'.
```
But a simple workaround exists: `generate if (1) begin : gen_generic`.

Reply:
I removed the generate block and only left the `if (1)` in line with our style guide to avoid generate/endgenerate. Let's see how DC likes that (should be OK, though).

Reply:
this works now